### ISSUE TITLE
External Keypad functionality

### ISSUE THREAD

**Issue by EssexPAL** [Denvi/Candle#318]
Hi

I'm sorry that this is a bit wordy but I hope that my explanation is reasonably clear.

Thanks for all the work on Candle which I use regularly (1.1.7) and find excellent, but ...

These observations are made from the Windows standpoint.

When I first started (using my CNC milling machine + Fusion 360) I rapidly discovered that what I desperately needed was a large, easily accessible, abort button.  I found that finding where the cursor was on the screen, then finding the Abort button and pressing it was much too slow culminating in a wrecked cutting tool.  The easiest route was to use an XKeys (XK-60) panel however 1.1.7 has very limited  keyboard accessibility so negating its value.

I therefore set out to write and app to send messages to the controls in Candle only to discover that it was written Qt so the components used are not the native Windows ones.  The next option was to write the code to emulate what a user does, move the mouse and press the mouse button.  This proved to be surprisingly successful however I came unstuck with the number of keyboard keys that the XKeys software has to emulate, there weren't really enough.  I therefore ended up taking the USB HID messages directly from the XKeys keypad thereby negating the need for any keyboard input at all, along with a heap more code this worked fine.  I then discovered 1.2.8b.

1.2.8b has shortcuts, the answer, well maybe.  The problem is that windows directly intercepts some of the avalable key sequences for carrying out OS related things, this lessens the available keys that can be used.  If you want to use the XKeys in hardware mode the number of keys available further restricted.  The other problem is if you switch to another app you really dont want the XKeys to deliver its keystrokes to any other app than Candle.  All in all it ends working but being messy.

Could I suggest a possible addition - That Candle responds to a pre-defined Message (from another app) and dependant upon the values of Wparam or Lparam calls the functionality of the appropriate button/control.  The values you use for short cuts currently could be kept the same and be delivered in WParam to be passed straight on.  This approach would still require an external application to configure the Xkeys buttons and originate the messages however it would leave the keyboard untouched and have no effect on any other app that happens to be running.  This would provide an external key pad interface without Candle needing to know anything about it.  I'm not suggesting that you remove shortcuts simply to add the messaging functionality.

Thanks

**Comment by ondras12345** [comment]
You can wire a Reset/Abort button directly to the machine (see [pinout](https://github.com/gnea/grbl/issues/123#issuecomment-279131776) for the common woodpecker control board)
